### ISSUE TITLE
Optimization of nearest-neighbor search in LSI

### ISSUE THREAD

**Issue by rth** [FreeDiscovery/FreeDiscovery#15]
For future reference.

Currently when computing document categorization with LSI, the nearst neighbours are found by computing all cosine similarities to relevant/non relevant documents. This may be significantly optimized using nearst-neighbour search in a pre-computed [kd-tree structure](http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KDTree.html) (with an euclidean metric), using the fact that 

          2*(1 - cosine_similarity) = euclidean_distance
for L2 normalized feature matrices.

**Edit:** actually the simplest would be to rewrite this nearest neigbour calculation in terms of 

     KNeighborsClassifier(n_neigbours=1, algorithm='kdtree')
